geoips/interfaces/module_based/readers.py

```diff
@@ -19,5 +24,233 @@ class ReadersInterface(BaseModuleInterface):
+              return dictionary keys).
+            * Read multiple resolutions of data if False.
+
+        Returns
+        -------
+        dict of xarray.Datasets
+            * dictionary of xarray.Dataset objects with required Variables and
+              Attributes.
+            * Dictionary keys can be any descriptive dataset ids.
+
+        See Also
+        --------
+        :ref:`xarray_standards`
+            Additional information regarding required attributes and variables
+            for GeoIPS-formatted xarray Datasets.
+        """
+        all_metadata = self.concatenate_metadata(
+            [call_single_file_func([x], metadata_only=True)["METADATA"] for x in fnames]
+        )
+        if metadata_only:
+            return all_metadata
+
+        dict_xarrays = self.call_files_and_get_top_level_metadata(
+            fnames,
+            all_metadata,
+            call_single_file_func,
+            metadata_only,
+            chans,
+            area_def,
+            self_register,
+        )
+        return dict_xarrays
+
+    def concatenate_metadata(self, all_metadata):
+        """Merge together metadata sourced from a list of files into one dictionary.
+
+        Where the structure of the merged metadata is a nested dictionary of metadata.
+
+        Ie. (xarray_obj has no data and is merely just a container for metadata):
+        {"METADATA": xobj.source_file_attributes: {fname: xobj, ..., "fnamex": xobj}}
+
+        Parameters
+        ----------
+        all_metadata: list of xarray.Datasets
+            - The incoming metadata from X number of files
+
+        Returns
+        -------
+        md: dict of xarray Datasets
+            - All metadata merged into a dictionary of xarray Datasets
+        """
+        md = {"METADATA": Dataset()}
+
+        for md_idx in range(len(all_metadata)):
+            # Set required attributes of the top-level metadata when this loop is
+            # started
+            if md_idx == 0:
+                md["METADATA"].attrs = all_metadata[md_idx].attrs
+                md["METADATA"].attrs["source_file_names"] = []
+                md["METADATA"].attrs["source_file_attributes"] = {}
+                md["METADATA"].attrs["source_file_datetimes"] = []
+                md["METADATA"].attrs["end_datetime"] = all_metadata[-1].end_datetime
+
+            # Add to optional attributes of the top-level metadata for each xobj
+            # provided
+            fname = all_metadata[md_idx].attrs["source_file_names"][0]
+            md["METADATA"].attrs["source_file_names"].append(basename(fname))
+            md["METADATA"].attrs["source_file_attributes"][basename(fname)] = (
+                all_metadata[md_idx]
+            )
+            md["METADATA"].attrs["source_file_datetimes"].append(
+                [
+                    all_metadata[md_idx].start_datetime,
+                    all_metadata[md_idx].end_datetime,
+                ],
+            )
```

Comment:
See here @mindyls.

Reply:
See above comment about source_file_datetimes, etc.  If that is what this conversation was referring to, please resolve this one @evrose54

Reply:
Resolving!